onnxruntime/python/session.py

```diff
@@ -192,9 +192,40 @@ def __init__(self, path_or_bytes, sess_options=None, providers=None, provider_op
         self._enable_fallback = True
         self._read_config_from_model = os.environ.get('ORT_LOAD_CONFIG_FROM_MODEL') == '1'
 
-        self._create_inference_session(providers, provider_options)
+        try:
+            self._create_inference_session(providers, provider_options)
+        except:
+            if self._enable_fallback:
+                # Collect fallback providers matching user's order
+                fallback_providers = []
+                fallback_providers_options = []
+                providers = providers or []
+
+                # Are there any user providers from the default fallback list?
+                for i, provider in enumerate(providers):
+                    if provider in self._fallback_providers:
+                        fallback_providers.append(provider)
+                        try:
```

Comment:
Do you need a try/catch here? Would ```if i < len(provider_options)``` not work?

Reply:
not sure if all this logic is really needed. it would be simpler if we just fall back to the hardcoded set of providers (same as the Run() path)
the goal of the fallback path is to activate a fail safe execution mode.
thus it's best to fall back to CPU if the session creation or inference run fails 
(because ORT guarantees CPU and CUDA provider should always works)
the other providers, not so much. 
incorporating user specified providers list would not provide such guarantees.

Reply:
@jywu-msft Given this check ```if provider in self._fallback_providers:``` I thought this method will always fallback to only the hardcoded set of providers. The objective of this logic is to preserve the provider options which get lost when we fallback in the run method.

Reply:
> @jywu-msft Given this check `if provider in self._fallback_providers:` I thought this method will always fallback to only the hardcoded set of providers. The objective of this logic is to preserve the provider options which get lost when we fallback in the run method.

that's true upon further inspection. i guess it's also up to the user to provide fallbacks in the correct order.
i wonder if we should separate out these cases. 
if the user provides the order and provider options, we'll go with the user provided information and treat it as an override.

If the user doesn't provide any explicit providers/options, then we fallback to the hardcoded set.

Reply:
The way I read this was - always use the hardcoded set of fallback providers, but if a user has provided any options (provider configs) for some of these fallback providers, use these options to construct the fallback providers rather than using the default set of options. In the run method, we completely ignore the user given provider configs for a particular provider when doing the fallback. 
I think the concept of a user providing fallback providers is separate and a new feature.

Reply:
Either way is fine with me, just thought to bring this issue(user's order) up for the consideration. Just let me know what is the common ground...

Reply:
@jywu-msft and I had a discussion. We think it's best to keep the behavior simple and consistent between run and session ctor. Hence, let's just use the hardcoded fallback providers.